TINY-4741: Fixed handling style attribute on styling elements

ltrouton opened the pull request:
This PR is to handle the case where a styling element has a style attribute e.g. 
`<p><strong class="abc" style="color: red;">abc</strong></p>`

The editor itself will never do this but external html may have this structure. Currently, when the bold toolbar is clicked to remove the bold, the colour is also removed, which from the user's perspective is incorrect.

I am thinking it might be worthwhile adding another option to the `remove` parameter for formats called `convert`. It essentially removes all attributes except the style attribute and converts the element to a span.

I think this is the safest approach as if styling elements which currently use `remove: 'all'`actually keep a style attribute in some cases, it technically goes against the definition of `remove: 'all'`.

Docs for reference:
https://www.tiny.cloud/docs/configure/content-formatting/#remove

This is a draft PR for now so happy to get feedback and make changes.

lnewson commented on modules/tinymce/src/core/main/ts/api/fmt/Format.ts:
```diff
@@ -55,15 +58,17 @@ export interface SelectorFormat extends CommonApplyFormat<SelectorFormat> {
 
 export interface CommonRemoveFormat<T> extends CommonFormat<T> {
-  remove?: 'none' | 'empty' | 'all';
+  remove?: 'none' | 'empty' | 'all' | 'convert';
```
I don't like the idea of this to be honest. The way I envisioned this would be solved, is to remove the current format rules and then check to see if it matched any other format. If it didn't then it should also remove the node.

lnewson replied:
This also potentially leads to spans without any styles, etc... being left around which isn't desirable.

lnewson replied:
Looking more, I can see why you went with the convert approach, though I still don't think it's a good option. `convert` will specifically only work for inline elements with styles such as this, so it's conflicting with the other types that work for everything (eg if the element had a format that used attributes instead this would still cause issues).

lnewson replied:
I'll take a better look over the weekend and see what other suggestions I can come up with.

ltrouton replied:
Thanks for having a look. I am more than happy to take a different approach. I think there are going to have to be some trade offs since it will be nearly impossible to cater to all of the different format possibilities. I'll see if I can think of a better solution that doesn't require adding new `remove` option.

lnewson replied:
To clarify I'm not against adding a new remove format, it's just if we do it needs to work for everything, not just styles+inline elements.

ltrouton commented:
I have been discussing with @lnewson on how best to handle this issue. We have decided that adding a new `preserve_attributes` parameter to inline formats is likely the safest option. It is not really intended to be used by users since it helps with an edge case so won't be documented.

I have also cleaned up `RemoveFormat.ts` a bit, removing some unnecessary let bindings and added types where applicable.

lnewson commented on modules/tinymce/src/core/main/ts/fmt/RemoveFormat.ts:
```diff
@@ -164,7 +159,7 @@ const removeNode = function (ed: Editor, node: Node, format) {
       if (parentNode === dom.getRoot()) {
         if (!format.list_block || !isEq(node, format.list_block)) {
-          each(Tools.grep(node.childNodes), function (node) {
+          Arr.each(Tools.grep(node.childNodes), (node) => {
```
Since you've swapped from `Tools.each` to `Arr.each` here, maybe also switch from `Tools.grep` to `Arr.filter`?

lnewson commented on modules/tinymce/src/core/main/ts/fmt/RemoveFormat.ts:
```diff
@@ -212,10 +207,29 @@ const removeFormat = function (ed: Editor, format: RemoveFormat, vars?: FormatVa
+        }
+      }
+    });
+    // Note: If there are no atrributes left, the element will be removed as normal at the end of the function
```
Typo here in attributes

- [ ] Fix typo

lnewson commented on modules/tinymce/src/core/main/ts/fmt/RemoveFormat.ts:
```diff
@@ -212,10 +207,29 @@ const removeFormat = function (ed: Editor, format: RemoveFormat, vars?: FormatVa
+    });
+    // Note: If there are no atrributes left, the element will be removed as normal at the end of the function
+    if (dom.getAttribs(elm).length > 0) {
+      // Conert inline element to span if necessary
```
Another typo here as well in convert

- [ ] Fix typo

lnewson commented on modules/tinymce/src/core/main/ts/fmt/RemoveFormat.ts:
```diff
@@ -212,10 +207,29 @@ const removeFormat = function (ed: Editor, format: RemoveFormat, vars?: FormatVa
   const elm = node as DomElement;
 
+  // Applies to styling elements like strong, em, i, u, etc. so that if they have styling attributes, the attributes can be kept but the styling element is removed
+  if (format.inline && Type.isArray(format.preserve_attributes)) {
```
Should this also have a `format.remove === 'all'` check? Or do we want to run this logic when doing `remove: 'empty'` and `remove: 'none'`?

lnewson replied:
Thinking about this more, we definitely don't want this to run on `remove: 'none'` as it won't be removing the element.

ltrouton replied:
Yeah, you are right. It makes most sense for it to only work with `format.remove === 'all'`. Will add that check in.

metricjs commented on modules/tinymce/src/core/main/ts/fmt/RemoveFormat.ts:
```diff
@@ -533,30 +544,33 @@ const remove = function (ed: Editor, name: string, vars?: FormatVars, node?: Nod
-          }, null, node);
-        }
+        // Remove parent span if it only contains text-decoration, yet a parent node also has the same text decoration.
+        const textDecorations = [ 'underline', 'line-through', 'overline' ];
```
If it's easy, could we add a test case for this too please? I know it's tangential, but would be nice.

ltrouton replied:
Added a test. Thought I should note that It appears a couple of tests fail when the text-decoration logic is altered/removed, namely `underline colors 6` and `underline colors 7` in FormatterApplyTest.ts so there should be enough coverage for this logic.

metricjs commented on modules/tinymce/changelog.txt:
```diff
@@ -1,9 +1,11 @@
 Version 5.3.0 (TBD)
     Added `uploadUri` and `blobInfo` to `editorUpload.uploadImages` return type #TINY-4579
+    Added `preserve_attributes` parameter to inline formats to assist with removing inline style element formats #TINY-4741
```
Hmm. I guess we should probably go with "if in doubt, don't document"?

ltrouton replied:
Yep. I agree it shouldn't be documented. That was my mistake for putting it in. Will remove it.
